todo_app/app.py

```diff
@@ -1,13 +1,85 @@
 import os
-from flask import Flask, render_template, request, redirect
-from todo_app.mongodbclient import *
-from todo_app.view import ViewModel
+from todo_app.models.view import ViewModel
 from todo_app.models.card import Card
+from todo_app.models.user import User
+
+from flask import Flask, render_template, request, redirect, url_for, session
+from todo_app.adapters.mongodb_todo import *
+from todo_app.adapters.mongodb_user import *
+
+#Login
+from flask_dance.contrib.github import make_github_blueprint, github
+from flask_login import UserMixin, current_user, LoginManager, login_required, login_user, logout_user
+from flask_dance.consumer import oauth_authorized
+
+class ReverseProxied(object):
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, environ, start_response):
+        scheme = environ.get('HTTP_X_FORWARDED_PROTO')
+        if scheme:
+            environ['wsgi.url_scheme'] = scheme
+        return self.app(environ, start_response)
+
+
+class UserToLogin (UserMixin):
+    def __init__(self, id):
+        self.id = id
 
 def create_app():
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
+    app.wsgi_app = ReverseProxied(app.wsgi_app)
+
+    #os.environ['OAUTHLIB_INSECURE_TRANSPORT']
+    gh_blueprint = make_github_blueprint(client_id=os.environ['GITHUB_CLIENT_ID'], client_secret=os.environ['GITHUB_CLIENT_SECRET'])
+    app.register_blueprint(gh_blueprint, url_prefix='/github_login')
+
+    todo = mongodb_todo()
+    usermanager = mongodb_user()
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.user_loader
+    def load_user(user_id):        
+        return UserToLogin(user_id)
+    
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(url_for('github_login'))
+    
+    @app.route('/logout')
+    @login_required
+    def logout():
```

Comment:
I know we talked this through already, but putting it in writing to point out that a logout endpoint doesn't mean a lot here - our app logs them out but if they return to the homepage then Github will just silently sign them back in! If we have this functionality, informing them they should also sign out of Github might make sense

Reply:
TY. I will look at this as part of my Mod 12.